Adding a JSON decoder for NdtResult objects

mtlynch opened the pull request:
Adds a JSON decoder so we can decode test results. Also adjusts the semantics
of the NdtResult & friends classes so they print better debug information
and support equality comparisons.

The client wrapper will not likely use this decoder directly, but it's
implemented here because consumers will need a way of decoding the JSON
output that the client wrapper produces and it seems easiest to have it
implemented once here rather than expect consumers to figure out how to
decode the JSON properly.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/ndt-e2e-clientworker/33)
<!-- Reviewable:end -->

pboothe commented:
Reviewed 1 of 4 files at r1.
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[client_wrapper/result_decoder.py, line 15 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33#-KFPavV0431T6rxVDufZ:-KFPavV0431T6rxVDuf_:-1226311114) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/a3038d68b92e5455de03c2f3eea85c87d080879d/client_wrapper/result_decoder.py#L15)):*
This assumes that the json is well formed and validated and has all the required fields and that the timestamp string is also well-formed. Maybe make a note of that?

---

*[client_wrapper/results.py, line 45 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33#-KFPcFdXlHtN7UEUyAXO:-KFPcFdYnypEVYRqsOj6:76628118) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/a3038d68b92e5455de03c2f3eea85c87d080879d/client_wrapper/results.py#L45)):*
Either make this actually JSON or don't make it look so much like JSON.

---

*[client_wrapper/results.py, line 160 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33#-KFPcS8_8jyXXB9PrnB4:-KFPcS8aoFYsRSzt80Uj:-1159500208) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/a3038d68b92e5455de03c2f3eea85c87d080879d/client_wrapper/results.py#L160)):*
Same JSON vs almost-JSON complaint here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
Review status: 1 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[client_wrapper/result_decoder.py, line 15 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33#-KFPavV0431T6rxVDufZ:-KFQMQJC8iIcHw2kjpuL:1466906601) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/a3038d68b92e5455de03c2f3eea85c87d080879d/client_wrapper/result_decoder.py#L15)):*
Good point, updated the docstring for `NdtResultDecoder`

Your note also forced me to check the documentation for `results.NdtResult` and I realized this PR changes the meaning of certain required fields, so I updated that docstring and updated tests/result_encoder.py to be consistent with this test. That's borderline out of scope, so I can pull it into a separate PR if you like, but I figured it was a pretty trivial edit.

---

*[client_wrapper/results.py, line 45 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33#-KFPcFdXlHtN7UEUyAXO:-KFQLgNCaGUSP9pcI6QN:-497468049) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/a3038d68b92e5455de03c2f3eea85c87d080879d/client_wrapper/results.py#L45)):*
Fixed

---

*[client_wrapper/results.py, line 160 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33#-KFPcS8_8jyXXB9PrnB4:-KFQLj8v39Wq4QNyMy4W:-497468049) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/a3038d68b92e5455de03c2f3eea85c87d080879d/client_wrapper/results.py#L160)):*
Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33)*
<!-- Sent from Reviewable.io -->

pboothe commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[client_wrapper/result_decoder.py, line 15 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33#-KFPavV0431T6rxVDufZ:-KFQPls3xxbuLgbIsWbA:-319017268) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/a3038d68b92e5455de03c2f3eea85c87d080879d/client_wrapper/result_decoder.py#L15)):*
Pulling it in is fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/33#-:-KFQQHSMYbeSTM2KWlTZ:1417088721)*
<!-- Sent from Reviewable.io -->
